File: kotlin/binary-search/main_1300.kt
import kotlin.math.min

fun main() {
    val n = readLine()!!.toInt()
    val k = readLine()!!.toInt()
    var start = 1
    var end = k
    var ans = 0
    while (start <= end) {
        val mid = (start + end) / 2
        var count = 0
        for (i in 1..n)
            count += min(mid / i, n)

        if (count < k) {
            start = mid + 1
        } else {
            ans = mid
            end = mid - 1
        }
    }
    println(ans)
}

File: cpp/step/binary-search/main_1300.cpp
#include <iostream>
#include <algorithm>
using namespace std;

int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(nullptr);
	cout.tie(nullptr);

	int n, k;
	cin >> n >> k;

	int start = 1;
	int end = k;
	int ans = 0;
	while (start <= end) {
		const int mid = (start + end) / 2;
		int sum = 0;
		for (int i = 1; i <= n; ++i) {
			sum += min(mid / i, n);
		}
		if (sum < k) {
			start = mid + 1;
		} else {
			ans = mid;
			end = mid - 1;
		}
	}
	cout << ans;
}
